// i/lib/Loader.js
var fs = require('fs');
var path = require('path');
var maker = require(I_ABS_PATH + '/lib/model/Maker.js').Maker;
var orms = require(APP_ABS_PATH + '/config/orm.js').orms;

var Loader = function() {};

Loader.prototype.loadCore = function loadCore() {
    // redis
    global.db = require('redis').createClient();

    global.I = {
        // common
        ConnectionPool: require('./ConnectionPool.js').ConnectionPool,
        Const: require('./Const.js').Const,
        DataPool: require('./DataPool.js').DataPool,
        Exception: require('./Exception.js').Exception,
        ExceptionCodes: require('./ExceptionCodes.js').ExceptionCodes,
        LogicController: require('./LogicController.js').LogicController,
        Route: require('./Route.js').Route,
        Syn: require('./Syn.js').Syn,
        Util: require('./Util.js').Util,
        // models
        Models: {
            PK: require('./model/PK.js').PK,
            Model: require('./model/Model.js').Model,
            List: require('./model/List.js').List,
            PKStore: require('./model/PKStore.js').PKStore,
            ModelStore: require('./model/ModelStore.js').ModelStore,
            ListStore: require('./model/ListStore.js').ListStore,
        },
    };
};

Loader.prototype.loadApp = function loadApp(appDir) {
    maker.createFiles(orms, APP_ABS_PATH + '/model');
    this.dir = appDir;

    this.loadDir('Ctrl', '/ctrl');
    this.loadDir('Const', '/const');
    this.loadDir('Logic', '/logic');
    this.loadEx();

    this.loadModelBaseClasses();
    this.loadDir('Models', '/model');
};

Loader.prototype.loadModelBaseClasses = function loadModelBaseClasses() {
    maker.makeModelBaseClasses(orms);
    this.loadMemoryModelClasses(maker.getClasses());
};

Loader.prototype.loadMemoryModelClasses = function loadMemoryClasses(classes) {
    for (var name in classes) {
        I.Models[name] = classes[name];
    }
};

Loader.prototype.loadClass = function loadClass(type, Class, name) {
    if (I[type] === undefined) I[type] = {};

    if (I[type][name]) throw new I.Exception(10003);
    I[type][name] = Class;
};

Loader.prototype.loadDir = function loadDir(type, dirname) {
    var self = this;
    var dir = this.dir + dirname;

    files = fs.readdirSync(dir);
    files.forEach(function(filename) {
        var name = path.basename(filename, '.js');
        var Class = require(dir + '/' + filename)[name];
        self.loadClass(type, Class, name);
    });
};

Loader.prototype.loadEx = function loadEx() {
    var exFullPath = this.dir + '/config/ex.js';
    var ex = require(exFullPath).ex;
    for (var i in ex) {
        I.ExceptionCodes[i] = ex[i];
    }
};

exports.Loader = new Loader();
